Fix #448 Process will not always exit when handler throws error

reisenberger opened the pull request:
Fix for #448 as outlined within discussion on that issue

micdenny commented:
Tomorrow I'll make some test, especially then ones described in the original post, if they green I will merge this and thanks you guys a lot :smile:

reisenberger commented:
@micdenny Yes, very happy if somebody else also tests it at the process level, like the original poster's #448 code.  

I was using the following slight amendment to the original poster's code:

```
namespace UnsafeShutdown
{
    class Program
    {
        private static Random _random = new Random();
        static void Main(string[] args)
        {
            Action<IServiceRegister> registerServices = register =>
            {
                register.Register<IEasyNetQLogger, ConsoleLogger>();
            };
            var bus = RabbitHutch.CreateBus("host=localhost", registerServices);

            bus.SubscribeAsync<TestPerformanceMessage>("foo", MessageHandler);

            bus.PublishAsync(new TestPerformanceMessage());
            Thread.Sleep(TimeSpan.FromSeconds(0.5));

            bus.Dispose();

            Thread.Sleep(TimeSpan.FromSeconds(5));

        }

        public static async Task MessageHandler(TestPerformanceMessage arg)
        {
            var delayUpToOneSecond = _random.Next(0, 1000);
            await Task.Delay(delayUpToOneSecond);
            Console.WriteLine("Throwing exception");
            throw new Exception("Handler failed");
        }
    }
}
```

The same in all essences, but to explain differences from the original poster's code:

+ Added `register.Register<IEasyNetQLogger, ConsoleLogger>();` to see what is going on!
+ I didn't at first see the original-poster's comment "run the EasyNetQ.Tests.Performance.Producer to get some events on the queue", so I added a `bus.PublishAsync(new TestPerformanceMessage());
`. The `Thread.Sleep(TimeSpan.FromSeconds(0.5));` after that is just a typical amount of time needed on my system to get the subscriber to pick up the message and throw its exception, before the consumer is disposed by `bus.Dispose()`.  I found this combination more convenient to repeatedly reproduce the issue, but one could also run the producer test as the OP suggests.
+ The 'Thread.Sleep(TimeSpan.FromSeconds(5));' on the end is only to give the chance to read the console messages if running pure exe, ie not in debug mode!

With the original (current) `DefaultConsumerErrorStrategy`, I could reproduce the bug often.  With the new, proposed `DefaultConsumerErrorStrategy` this issue was removed.

reisenberger commented:
@micdenny At https://github.com/EasyNetQ/EasyNetQ/commit/212744705083bf89b36e0a4bd6d44d18a6cb142c 
made the code tighter again, by placing all modifications to the `connection` variable within locks.  This eliminates a further possible race condition I observed in step-debugging.

micdenny commented on Source/EasyNetQ/Consumer/DefaultConsumerErrorStrategy.cs:
```diff
@@ -200,12 +210,20 @@ private string CreateConnectionCheckMessage()
+            disposing = true;
 
-            if(connection != null) connection.Dispose();
+            lock (syncLock)
```
actually I don't like the idea of having locks in dispose, could we discuss a little bit more on this?

micdenny replied:
probably we can't do so many different things if we want to create a pure thread-safe disposable class

reisenberger replied:
@micdenny true, lock in Dispose is probably bad: it could block the finalizer thread.  

It was added to avoid a complex (observed in debugging) race condition between `Dispose()` and `Connect()` where `if(connection != null) connection.Dispose();` did nothing (because connection was still null), but the `Connect()` method had already passed the new `&& !(disposing || disposed)` condition, and so continued to make a new `connection = connectionFactory.CreateConnection();` (this connection never got disposed and still held the application open).  The lock avoided this, but I agree lock in Dispose seems bad.  This particular race condition can probably be fixed another way.  I will propose an alternative shortly, which removes the lock from Dispose() method.

micdenny replied:
> I will propose an alternative shortly, which removes the lock from Dispose() method.

great :+1:

reisenberger replied:
Now added commit which removes the lock from Dispose() method.

reisenberger commented:
@micdenny Great, will leave you to merge when a suitable time.

If anyone is still having problems after this is merged, with executables not closing due to extra RabbitMQ connection threads open with connections ... (for example @MichaelLogutov mentioned in #448 he had maybe this problem without using `DefaultConsumerErrorStrategy`) ... maybe we could look whether `PersistentConnection` or `ConnectionFactoryWrapper` have any similar race conditions during disposal - or if they are continuing to re-connect _after_ disposal.  I had a brief look at `PersistentConnection` : there is some good defence against using after has been disposed, but it could be stronger, eg in `TryToConnect()` method. ... But this is somewhat speculation, based on issues seen in `DefaultConsumerErrorStrategy`...

reisenberger commented:
@micdenny `PersistentConnection.TryToConnect(...)` could be improved similarly in these two simple ways:

```
do
{
     // same as before
} while (!disposed && connectionFactory.Next());
```
And near the end:
```
if (!disposed) { StartTryToConnect(); }
```

Let me know if you would like this added.  We don't have a _proven_ problem in `PersistentConnection`, but this would provide stronger do-not-reconnect-after-disposed logic, similar to what we have just done for `DefaultConsumerErrorStrategy`

reisenberger commented:
@micdenny It would be good to get this merged as well (after the higher-profile merges such as async implementation #461).

Regarding the fact that we seem to have different kinds of application-hang scenario:
(a) #448 related to the `DefaultConsumerErrorStrategy` (fixed by this PR)
(b) other hangs, unrelated to the `DefaultConsumerErrorStrategy` 

I suggest we merge this #500 first to be sure we have dealt with (a), then tackle (b) separately.

micdenny commented:
> PersistentConnection.TryToConnect(...) could be improved similarly in these two simple ways:
> Let me know if you would like this added

I like also this one if you want to add, thanks for all this job!

> I suggest we merge this #500 first to be sure we have dealt with (a), then tackle (b) separately.

deal!

reisenberger commented:
Happy to rebase this for next-merge, after #461 and #498.  (Will await those merges...)

micdenny commented:
> Happy to rebase this for next-merge

thanks!

reisenberger commented:
@micdenny I am mystified as to why PR #500 says "This branch has conflicts that must be resolved".  Git shell locally on my machine, reports no conflicts.  I cannot find any conflicts in my local file set (though file AdvancedApiPingPongTests contains lots of >>>>> <<<< that maybe looks like a conflict).  

AppVeyor reports successful check. Is this ready now for merge?

reisenberger commented:
-:( AppVeyor reports sometimes successful check, sometimes not ??

reisenberger commented:
@micdenny Ok, now re-based properly.  (Seems I was just too hasty with my re-basing commit earlier ...)

This ready to merge, or happy if this is after #498.  Will work on 

>(b) other hangs, unrelated to the  DefaultConsumerErrorStrategy  

... after this merged.

micdenny commented:
@reisenberger I did merge #498 can you bumb the version again please :smile:

reisenberger commented:
@micdenny Now rebased and bumped version.cs.  

After this is merged, will take a new branch and work on:

>(b) other hangs, unrelated to the DefaultConsumerErrorStrategy

zidad commented:
@reisenberger since there are quite some commits but only a small amount of changes, do you think it's a good idea to squash all commits into 1 to make it a bit more clear which commits belong to this feature?

micdenny commented:
@zidad [I did it too](https://github.com/micdenny/EasyNetQ/tree/reisenberger-consumererrorstrategyhangfix) just to try the git interactive feature, that's cool :smile: :+1: 

the only strange thing is this, but that's not a big deal:

![image](https://cloud.githubusercontent.com/assets/1633595/11468278/8abc1806-974e-11e5-91b3-75795de87dde.png)

reisenberger commented:
@zidad, @micdenny  Very happy if we simply instead start a new pull-request from the squashed commit either of you have created (or, do you want me to recreate ths?).  However, the commit comments in your merged version also contain comments about the (dropped-for-the-moment) SslOptions #477 proposal, which I initially had in the same branch, and so then reverted out.  We would want to remove these comments from the squashed commit message, before merging this pull request:

>Allow config SslOptions from string via CreateBus
>Adds an SslOptionsStringParser to allow SslOptions to be configured from
>a string (eg drawn from a config file) rather than having to hand-craft
>SslOption and ConnectionConfiguration objects.
>Add RabbitHutch.CreateBus(...) overloads with sslOptionsString
>parameter.
>Revert "Allow config SslOptions from string via CreateBus"
>This reverts commit 0066606.

micdenny commented:
for the reason of reverting commits, would be definitively better to squash commits and yes removing that comment, I can do it and open a new PR and as @zidad said, you remain as author and me just as committer, so we know who point the finger to :joy: :joy:

zidad commented:
@micdenny yes that can really tidy up your history. There's no real use in keeping all the intermediate commits that might be broken anyway.

@reisenberger I can change the commit to whatever you like, if you like to use your branch I don't think you need to start a new pull request. 

You could also try the rebasing on this branch yourself by just:
- running `git rebase master -i`, 
- pick the first commit, replace `pick` on the other lines with `s`
- cycle through the conflicts using `git mergetool` and `git rebase --continue` for as long as necessary
- in the end you'll end up with 1 commit, you can still amend.
- You'll have to use `push -f` to push it to the server....

micdenny commented:
+1 for @zidad steps @reisenberger tell us if you can do it your self or need help on this

reisenberger commented:
Thanks @zidad for steps - corresponds with what I was looking up on github help/google too.  Useful feature to know.  Will try this later in day.

reisenberger commented:
@micdenny @zidad What I attempted here was to group-and-squash the commits, so as to end up with two commits: one (internal order preserved) encompassing all the rebasing (squashed together) and another (internal order preserved) encompassing all the #448 changes (squashed together).  The two streams of changes should have been orthogonal so this shouldn't have presented a problem (I do this all the time in SVN), however, the result here still wasn't/isn't very clean.  

At this point I think the most effective use of time would be to take and merge down the squashed version either of you @micdenny @zidad have created.

zidad commented:
Ok! https://github.com/EasyNetQ/EasyNetQ/pull/510

reisenberger commented:
Closed via #510
